### PR TITLE
Generalise kron to an arbitrary number of dimensions.

### PR DESCRIPTION
**Pull request by hameerabbasi**
Generalizes `kron` to an arbitrary number of dimensions.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
cc @jcrist for review.

File: sparse/tests/test_coo.py [hunk 1]
```diff
@@ -334,10 +334,10 @@ def test_kron(a_ndim, b_ndim):
         assert_eq(sparse.kron(a, b), sol)
 
 
-@pytest.mark.parametrize('ndim', [0, 1, 2])
```
Comment (hameerabbasi) [inline sparse/tests/test_coo.py]: This was, again, a case better handled in the dense regime, so I removed it.

File: sparse/tests/test_coo.py [hunk 2]
```diff
@@ -314,11 +314,11 @@ def test_dot_nocoercion():
...
 
-@pytest.mark.parametrize('a_ndim', [1, 2])
-@pytest.mark.parametrize('b_ndim', [1, 2])
+@pytest.mark.parametrize('a_ndim', [1, 2, 3])
+@pytest.mark.parametrize('b_ndim', [1, 2, 3])
 def test_kron(a_ndim, b_ndim):
-    a_shape = (10, 11)[:a_ndim]
-    b_shape = (12, 13)[:b_ndim]
+    a_shape = (3, 4, 5)[:a_ndim]
+    b_shape = (4, 5, 3)[:b_ndim]
```
Comment (jcrist) [inline sparse/tests/test_coo.py]: It'd be maybe good to test:
- Using dimensions that don't ever match up (`len(set(a_shape).intersection(b_shape)) == 0`) to ensure no accidental reliance on broadcasting aligning. Perhaps `(3, 4, 5)`, and `(6, 7, 8)`.
- Cases with a nonzero fill value

File: sparse/coo/common.py [hunk 3]
```diff
@@ -244,44 +240,38 @@ def kron(a, b):
...
     """
-    from .core import COO, as_coo
-    check_zero_fill_value(a, b)
+    from .core import COO
+    from .umath import _cartesian_product
 
-    a_sparse = isinstance(a, SparseArray)
-    b_sparse = isinstance(b, SparseArray)
+    a_sparse = isinstance(a, (SparseArray, scipy.sparse.spmatrix))
+    b_sparse = isinstance(b, (SparseArray, scipy.sparse.spmatrix))
```
Comment (jcrist) [inline sparse/coo/common.py]: We don't ever test with a `scipy.sparse.spmatrix`, is this something that should also be added to the tests?

File: sparse/tests/test_coo.py (hunk 2 above)
Comment (jcrist) [reply sparse/tests/test_coo.py]: The first one might be being overly paranoid though, feel free to ignore.

Comment (hameerabbasi) [reply sparse/tests/test_coo.py]: Ah, yes. My algorithm only works with a zero fill-value.

File: sparse/coo/common.py (hunk 3 above)
Comment (hameerabbasi) [reply sparse/coo/common.py]: Added a test.

Comment (hameerabbasi) [reply sparse/coo/common.py]: For future references, `asCOO` and the `COO` constructor are pretty heavily tested, so we probably don't need to test things like these everywhere.

**Pull request merged**